### PR TITLE
add zero-cost accuracy predictor metrics

### PR DESCRIPTION
**Pull request by lzrwch**
- [x] functional tests (dataloader, generator)
- [x] add `gmacs` and `nparams` via `profile` method

### CODE REVIEW

**Comment by lzrwch** [comment]
@matteog11 I have cleaned up the code a little bit and changed the interface. One gets the zero cost estimator via 
```
from deeplite_torch_zoo import get_zero_cost_estimator

estimator_fn = get_zero_cost_estimator('snip')
```
Each estimator function takes the following args: 1. `model`, 2. `loss_fn` (not required for some estimators), and 3. either `dataloader` or, if customization is required, a `model_output_generator`, a generator that returns a tuple `inputs, outputs, targets, loss_kwargs`. `loss_fn` must be a method that takes the following args `outputs, targets` and also optionally `loss_kwargs` unpacked to kwargs. 

A simple case when passing `dataloader` is enough:
```
import torch
import torch.nn as nn

from deeplite_torch_zoo import get_model, get_dataloaders, get_zero_cost_estimator
from deeplite_torch_zoo.src.registries import ZERO_COST_SCORES

model = get_model(model_name='resnet18', dataset_name='cifar100')
dataloaders = get_dataloaders(data_root='./', model_name='resnet18', dataset_name='cifar100', batch_size=64, num_workers=8)
loss = nn.CrossEntropyLoss()

for metric_name in ZERO_COST_SCORES.registry_dict:
    zero_cost_estimator = get_zero_cost_estimator(metric_name)
    res = zero_cost_estimator(model, dataloader=dataloaders['test'], loss_fn=loss)
    print(metric_name, res)
```

Example with customization for YOLO (taken from zoo develop) via passing `model_output_generator`: 
```
import torch
from deeplite_torch_zoo import get_model, get_dataloaders, get_zero_cost_estimator
from deeplite_torch_zoo.src.registries import ZERO_COST_SCORES

from deeplite_torch_zoo.src.object_detection.yolov5.losses.yolov5_loss import YoloV5Loss
from itertools import repeat

IMG_SIZE = 320
model = get_model(model_name='yolo5n', dataset_name='voc', pretrained=False)
dataloaders = get_dataloaders(
    data_root='/neutrino/datasets/VOCdevkit', 
    model_name='yolo5n', 
    dataset_name='voc', 
    batch_size=64, 
    num_workers=8,
    img_size=IMG_SIZE
)

criterion = YoloV5Loss(model, device='cpu')
dataloader = dataloaders['test']


def model_output_generator(model, shuffle_data=True, input_gradient=False):
    loader = dataloader if shuffle_data else repeat(next(iter(dataloader)))
    for imgs, targets, labels_length, _ in loader:
        imgs.requires_grad_(input_gradient)
        pred = model(imgs)  # forward
        yield imgs, pred, targets, {'labels_length': labels_length}


def compute_loss_fn(outputs, targets, labels_length=None):
    loss, _ = criterion(
        outputs, targets, labels_length, IMG_SIZE
    )
    return loss


def yolo_post_processing(outputs):
    return torch.cat([o.view(outputs[0].shape[0], 3, -1) for o in outputs], 2)


for metric_name in ZERO_COST_SCORES.registry_dict:
    zero_cost_estimator = get_zero_cost_estimator(metric_name)
    kwargs = {
        'output_post_processing': yolo_post_processing
    } if metric_name == 'jacob_cov' else {}
    res = zero_cost_estimator(
        model, 
        model_output_generator=model_output_generator, 
        loss_fn=compute_loss_fn,
        **kwargs,
    )
    print(metric_name, res)
```
One extra piece of customization here that is required only for `jacob_cov` metric is one must pass an `output_post_processing` method that concats the 3 output tensors from YOLO to one for gradient computation.

**Comment by lzrwch** [comment]
NB: `gmacs` and `nparams` are to be added. Not adding `zen_score` for now.

File: deeplite_torch_zoo/src/zero_cost_proxies/grasp.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+@ZERO_COST_SCORES.register('grasp')
+def grasp(model, model_output_generator, loss_fn, mode='param', T=1, niter=1):
+    weights = []
+    for module in model.modules():
+        if isinstance(module, nn.Conv2d) or isinstance(module, nn.Linear):
+            weights.append(module.weight)
+            module.weight.requires_grad_(True)
+
+    # Forward n1
+    data_generator = model_output_generator(model, shuffle_data=False)
```
Comment (lzrwch) [inline deeplite_torch_zoo/src/zero_cost_proxies/grasp.py:17]: @matteog11 `shuffle_data` arg is required here, since I assume this metric requires the same batch of data for all forward passes

**Comment by lzrwch** [comment]
@matteog11 

fisher 0.411
grad_norm 0.258
grasp 0.038
jacob_cov 0.204
l2_norm 0.508
plain 0.075
snip 0.471
synflow 0.514
zico 0.465
nparams 0.622
macs 0.610

| ZC metric   |      Kendall-Tau      |
|----------|:-------------:|
| fisher |  left-aligned |
| grad_norm |    centered   |
| grasp | right-aligned |
| grasp | right-aligned |
| grasp | right-aligned |
| grasp | right-aligned |
| grasp | right-aligned |
| grasp | right-aligned |

**Comment by lzrwch** [comment]
@matteog11 here are the results obtained with this implementation:

500 randomly sampled models from NASBench201 (CIFAR100):

| ZC metric   |      Kendall-Tau    |
|----------|:-------------:|
| fisher | 0.354 |
| grad_norm |    0.301   |
| grasp | 0.307 |
| jacob_cov | 0.561 |
| l2_norm | 0.499 |
| plain | -0.190 |
| snip | 0.412 |
| synflow | 0.548 |
| zico | 0.538 |
| nparams | 0.531 |
| macs | 0.567 |

PyTorchCV ImageNet models:

| ZC metric   |      Kendall-Tau    |
|----------|:-------------:|
| fisher | 0.411 |
| grad_norm |    0.258   |
| grasp | 0.038 |
| jacob_cov | 0.204 |
| l2_norm | 0.508 |
| plain | 0.075 |
| snip | 0.471 |
| synflow | 0.514 |
| zico | 0.465 |
| nparams | 0.622 |
| macs | 0.610 |

I think it's ready to be merged now, please have a look and let me know

File: deeplite_torch_zoo/src/zero_cost_proxies/grasp.py (hunk 1 above)
Comment (lzrwch) [reply deeplite_torch_zoo/src/zero_cost_proxies/grasp.py:33]: @matteog11 to confirm

File: deeplite_torch_zoo/src/classification/augmentations/augs/augs.py [hunk 2]
```diff
@@ -29,6 +29,7 @@ def get_imagenet_transforms(
     re_prob=0.0,  # Random erase prob (default: 0.)
     re_mode='pixel',  # Random erase mode (default: "pixel")
     re_count=1,  # Random erase count (default: 1)
+    **kwargs,
```
Comment (alexh-deeplite) [inline deeplite_torch_zoo/src/classification/augmentations/augs/augs.py:32]: is this meant to be a part of ZC PR?

**Review by alexh-deeplite — APPROVED** [review]

File: deeplite_torch_zoo/api/eval/zero_cost.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                'a standard interface to compute model output is assumed.'
+            )
+
+        model_ = model
+        if not inplace:
+            model_ = deepcopy(model)
+        device = next(model.parameters()).device
+
+        if model_output_generator is None:
+            def model_output_generator(model, shuffle_data=True, input_gradient=False):
```
[the remark this answers is not in the document]
Comment (lzrwch) [reply deeplite_torch_zoo/api/eval/zero_cost.py]: agreed, done

File: deeplite_torch_zoo/src/classification/augmentations/augs/augs.py (hunk 2 above)
Comment (lzrwch) [reply deeplite_torch_zoo/src/classification/augmentations/augs/augs.py]: removed

**Review by hossein-dl — APPROVED** [review]
LGTM

File: deeplite_torch_zoo/src/zero_cost_proxies/synflow.py
[the remark this answers is not in the document]
Comment (lzrwch) [reply deeplite_torch_zoo/src/zero_cost_proxies/synflow.py:1]: originally we did not add the `dummify_bns` argument to synflow calculation. it dynamically replaces forward methods in normalization layers, making them into identity layers. another approach can be setting `dummify_bns=False` and `bn_training_mode=True`, we found in that case correlation with accuracy to also be on a similar level

**Comment by matteog11** [comment]
The general code looks good to me, I just have a few minor comments:

- [This part](https://github.com/Deeplite/deeplite-torch-zoo/blob/32f52f4677b5bb5b0e165774b5f7c11ff01cfb95/deeplite_torch_zoo/api/eval/zero_cost.py#L70) could be edited to `model_.eval() if eval_mode else model_.train()`.
            
- We may remove the duplicate `model.zero_grad()` inside each score function (e.g. [here](https://github.com/Deeplite/deeplite-torch-zoo/blob/32f52f4677b5bb5b0e165774b5f7c11ff01cfb95/deeplite_torch_zoo/src/zero_cost_proxies/snip.py#L54)), considering that it is already applied [before](https://github.com/Deeplite/deeplite-torch-zoo/blob/32f52f4677b5bb5b0e165774b5f7c11ff01cfb95/deeplite_torch_zoo/api/eval/zero_cost.py#L71) calling `compute_zc_score_fn()`.

File: deeplite_torch_zoo/src/zero_cost_proxies/grasp.py (hunk 1 above)
Comment (matteog11) [reply deeplite_torch_zoo/src/zero_cost_proxies/grasp.py:33]: Some references: 
- Algorithm from the original `GraSP` [paper](Picking Winning Tickets Before Training by Preserving Gradient Flow)

Screenshot 2023-06-28 at 18 50 55

- [NASLib](https://github.com/automl/NASLib/blob/172f27f773f49c35467423b8e3e1c203271b2a7a/naslib/predictors/pruners/measures/grasp.py#L45) code for computing `GraSP` score.

Comment (matteog11) [reply deeplite_torch_zoo/src/zero_cost_proxies/grasp.py:33]: My understanding is that the batch should be the same across the 2 different forward passes. [This](https://github.com/Deeplite/yolobench/blob/4530e800e0f33f399e0c4665a3ba5c52199dc915/zero_cost/measures/grasp.py#L30C21-L30C27) is how I implemented it in the previous version of the code. However, please double-check the references I shared above as I am not completely sure about it.
